backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/CodeGenerator.kt

```diff
@@ -625,28 +625,46 @@ internal class FunctionGenerationContext(val function: LLVMValueRef,
         appendingTo(lpBlock) {
             val landingpad = gxxLandingpad(2)
             LLVMAddClause(landingpad, kotlinExceptionRtti.llvm)
+            LLVMAddClause(landingpad, objcNSExceptionRtti.llvm)
             LLVMAddClause(landingpad, LLVMConstNull(kInt8Ptr))
 
+            val foreignExceptionBlock = basicBlock("foreignException", position()?.start)
             val fatalForeignExceptionBlock = basicBlock("fatalForeignException", position()?.start)
+            val forwardNativeExceptionBlock = basicBlock("forwardNativeException", position()?.start)
             val forwardKotlinExceptionBlock = basicBlock("forwardKotlinException", position()?.start)
 
+            val typeId = extractValue(landingpad, 1)
             val isKotlinException = icmpEq(
-                    extractValue(landingpad, 1),
+                    typeId,
                     call(context.llvm.llvmEhTypeidFor, listOf(kotlinExceptionRtti.llvm))
             )
 
-            condBr(isKotlinException, forwardKotlinExceptionBlock, fatalForeignExceptionBlock)
+            condBr(isKotlinException, forwardKotlinExceptionBlock, foreignExceptionBlock)
+
+            appendingTo(foreignExceptionBlock) {
+                val isObjCException = icmpEq(
+                        typeId,
+                        call(context.llvm.llvmEhTypeidFor, listOf(objcNSExceptionRtti.llvm))
+                )
+                condBr(isObjCException, forwardNativeExceptionBlock, fatalForeignExceptionBlock)
+            }
 
             appendingTo(forwardKotlinExceptionBlock) {
                 // Rethrow Kotlin exception to real handler.
                 codeContext.genThrow(extractKotlinException(landingpad))
             }
 
+            appendingTo(forwardNativeExceptionBlock) {
```

Comment:
This part doesn't make sense (and actually doesn't work) for non-Apple platforms yet.